### DISCUSSION TITLE
AssertionError with example for heterogeneous GNN

### DISCUSSION THREAD

**Discussion by Zytrus** [pyg-team/pytorch_geometric#4485] [Q&A] [answered]
Hello,

I have encountered a mysterios error while trying to execute a graph classification with one of the examples given in [Heterogeneous GNN - Automatically Converting GNN Models (GAT)](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#automatically-converting-gnn-models). 
I specifically tried out the GAT GNN:
```py
from torch_geometric.nn import GATConv, Linear, to_hetero

class GAT(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GATConv((-1, -1), hidden_channels, add_self_loops=False)
        self.lin1 = Linear(-1, hidden_channels)
        self.conv2 = GATConv((-1, -1), out_channels, add_self_loops=False)
        self.lin2 = Linear(-1, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index) + self.lin1(x)
        x = x.relu()
        x = self.conv2(x, edge_index) + self.lin2(x)
        return x


model = GAT(hidden_channels=64, out_channels=dataset.num_classes)
model = to_hetero(model, data.metadata(), aggr='sum')
```
and am training it like this:
```py
def train():
    model.train()

    for data_train in train_dataset:  # Iterate in batches over the training dataset.
        out = model(data_train.x_dict, data_train.edge_index_dict)  # Perform a single forward pass.
        loss = criterion(out, data_train.y)  # Compute the loss.
        loss.backward()  # Derive gradients.
        optimizer.step()  # Update parameters based on gradients.
        optimizer.zero_grad()  # Clear gradients.

def test(loader):
    model.eval()

    correct = 0
    for data_test in loader:  # Iterate in batches over the training/test dataset.
        out = model(data_test.x_dict, data_test.edge_index_dict)  
        pred = out.argmax(dim=1)  # Use the class with highest probability.
        correct += int((pred == data_test.y).sum())  # Check against ground-truth labels.
    return correct / len(loader)  # Derive ratio of correct predictions.

for epoch in range(1, 15):
    train()
    train_acc = test(train_dataset)
    test_acc = test(test_dataset)
    print(f'Epoch: {epoch:03d}, Train Acc: {train_acc:.4f}, Test Acc: {test_acc:.4f}')
```
Since my dataset consists of only 29 graphs, I don't need to batch the graphs, but I get this mysterious error:
```
Traceback (most recent call last):
  File "C:\Users\Documents\Privacy\test.py", line 89, in <module>
    train()
  File "C:\Users\Documents\Privacy\test.py", line 71, in train
    out = model(data_train.x_dict, data_train.edge_index_dict)  # Perform a single forward pass.      
  File "C:\Users\miniconda3\lib\site-packages\torch\fx\graph_module.py", line 616, in wrapped_call
    raise e.with_traceback(None)
AssertionError
```
I am genuinely confused as to why the ```wrapped_call(self, *args, **kwargs)``` would throw an Exception, because it does not really give a hint, on where the mistake could be...

Help would be greatly appreciated!

Best regards, Zytrus

**Comment by rusty1s** [comment, marked as answer]
Thanks for reporting and sorry for this mysterious error message. My assumption is that the metadata does not match across all your graphs (which is a requirement). That means, the following should pass:
```python
for data in dataset:
    assert data.metadata() == dataset[0].metadata()
```

**Reply by Zytrus** [reply]
Hello rusty1s,

I have inserted the code you suggested, and it runs through for me. It stops at the same spot as before and the ```wrapped_call(self, *args, **kwargs)``` is still throwing an exception...

Kind regards, Zytrus

**Reply by Zytrus** [reply]
I have now copied the HGT GNN from [Deploy Existing Heterogenous Operators](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#deploy-existing-heterogenous-operators) and had to change edge_index dtype to torch.long in order for it to run through. The model doesn't throw an error anymore, but now the loss function throws:
```
Traceback (most recent call last):
  File "C:\Users\Documents\Privacy\test.py", line 99, in <module>
    train()
  File "C:\Users\Documents\Privacy\test.py", line 82, in train
    loss = criterion(out, data_train.y)  # Compute the loss.
  File "C:\Users\miniconda3\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\miniconda3\lib\site-packages\torch\nn\modules\loss.py", line 1150, in forward
    return F.cross_entropy(input, target, weight=self.weight,
  File "C:\Users\miniconda3\lib\site-packages\torch\nn\functional.py", line 2846, in cross_entropy
    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
ValueError: Expected input batch_size (7) to match target batch_size (1).
```
Even though I don't use batches for my training it is complaining about batches... is it necerssary to batch the dataset?

Kind regards, Zytrus

**Reply by rusty1s** [reply]
How do the input shapes of prediction and ground-truth labels look like in the loss function? These need to match.

**Reply by Zytrus** [reply]
Ah yes, I see my mistake, the model returns a dictionary of the predicted values for each type, but I need a classification between 0 and 1. Is hierarchical pooling possible with PyG? I remember reading that a lot of information gets lost with naive pooling at the end..

and another issue I bumped into, is that my values are huge. And the calculated accuracy is always above 2... how do I tackle this issue? Does the dataset need to be normalized?

And do you perhaps know a public code base for heterogeneous GNN doing graph classification? I feel like I probably need some orientation in order to get on the right track, because the majority of code I find online do node-level classification...

Kind regards, Zytrus

**Reply by rusty1s** [reply]
I am not aware of any publicly available heterogenous graph classification dataset, so we do not have an example of this yet. Sorry. In general, I think you should go with global pooling before trying out any fancier hierarchical pooling variant. In practice, these usually perform equally well. For global pooling, you would need to pool each type separately, and either sum or concatenate the result:
```python
outs = []
for x, batch in zip(x_dict.values(), batch_dict.values()):
    outs.append(global_mean_pool(x, batch)
out = torch.cat(outs, dim=-1)
```

**Reply by Zytrus** [reply]
Thank you for the fast reply! Is ```batch_dict.values()``` necessary? Since I am not working with batches, is it okay to just leave it out or would that cause bad results?
I'm just a little confused, because an error message appears, saying it requires the batch argument, but in the documentation it says ```batch (LongTensor, optional)```, so is it optional?

EDIT: Okay so I've just implemented DataLoaders, and the code runs.. although I'm not sure how and why it's working, it's somehow working. 
Now to an maybe amateur question, after some trial and error I found that the predictions were not classifying between 0 and 1 at first, but classifying classes like 20 or 80 (which don't exist). Because of the training they eventually start looking like the ground truth, is this normal? Are they allowed to classify between classes that don't even exist?

Oh and the accuracy I get is somehow above 1, usually around 4.... which is probably not quite right...
```python
def test(loader):
    model.eval()

    correct = 0
    for data_test in loader:  # Iterate in batches over the training/test dataset.
        out = model(data_test.x_dict, data_test.edge_index_dict, data_test.batch_dict)  
        # pred = out['process'].argmax(dim=1)  # Use the class with highest probability.
        pred = out.argmax(dim=1)  # Use the class with highest probability.
        correct += int((pred == data_test.y).sum())  # Check against ground-truth labels.
    return correct / len(loader)  # Derive ratio of correct predictions.
```
The only possibility would be```correct > len(loader)```, so there are more correct labels than there are elements in the loader. This could be due to the batching process, it splits the data and even though there was a clear y=0/1 for each HeteroData, now y = [num_graphs_in_batch] for each graph. From what I understood the batches kind of combine all given graphs into a big batch... so I'm wondering how the accuracy can be calculated correctly with batches?

And if we're working with batches... Say we deploy this GNN, and it gets a new heterogeneous graph as an input. Since it has only trained on batches before, how can it even handle single graphs?
What would the prediction for that look like? 

Sorry for the many questions, but this does pique my interest, because I have come to the point, where I don't fully understand my code anymore....

Kind regards, Zytrus

**Reply by rusty1s** [reply]
batch_dict is not required if you are operating on single graphs. Support for optional batches just got recently added to PyG and is not yet released. As a workaround, you can also simply call x.mean(dim=-1, keepdim=True) for batches of size 1.

However, it looks like you make use of batches though. Note that you need to divide by len(loader.dataset) instead.

**Reply by Zytrus** [reply]
Yes thank you, I've noticed that mistake in my evaluation as well! I'm working with batches now and it runs through!
But is it normal the classification doesn't classify between the two classes 1 and 0 at first?

**Reply by rusty1s** [reply]
This shouldn't be normal. Is the `out_channels` set correctly in your case? You would want your neural network to only output two channels in this case.

**Reply by Zytrus** [reply]
Hello rusty1s,

thank you for the advice, I did forget to put a linear layer at the end of the model, so the output wasn't only between 2 classes. But after adding the linear layer defined as ```self.lin = Linear(hidden_channels, out_channels)```, the output still wasn't between 0 and 1.

Maybe it has something to do with how I get the final values in the forward function?

```python
for x, batch in zip(x_dict.values(), batch_dict.values()):
            if biggest_size < torch.max(batch):
                biggest_size = torch.max(batch) + 1
            x = global_mean_pool(x, batch, size=biggest_size)
            x = F.dropout(x, p=0.5, training=self.training)
            outs.append(self.lin(x))

        out = torch.cat(outs, dim=-1)
```

Sorry for the late reply, and thank you for the help so far!

Kind regards, Zytrus

**Reply by rusty1s** [reply]
If you want to enforce outputs between `0` and `1`, you should be able to enforce this via a simple `sigmoid`:
```python
out = torch.cat(outs, dim=-1).sigmoid()
```
Note that this is usually not necessary for PyTorch loss functions though.

**Reply by Zytrus** [reply]
Ah okay, so the results are normal?
And is it normal, that the model performs worse with the dropout and linear layer? Sometimes it performs better, sometimes worse, because my dataset shuffles everytime the model runs through.

That means, that the model isn't actually optimal for the dataset I'm using, correct?

Kind regards,
Zytrus

**Reply by rusty1s** [reply]
This is hard to tell. Can you tell me more about how you apply loss function and evaluation metrics?

**Reply by Zytrus** [reply]
I used the training from the example given in the documentation, although adjusted it a little:
```python
criterion = torch.nn.CrossEntropyLoss()
def train():
    model.train()

    total_examples = total_loss = 0
    for batch in train_loader:
        optimizer.zero_grad()
        # batch = batch.to('cuda:0')
        batch_size = batch.num_graphs
        out = model(batch.x_dict, batch.edge_index_dict, batch.batch_dict)
        loss = criterion(out[:batch_size], batch.y[:batch_size])
        loss.backward()
        optimizer.step()

        total_examples += batch_size
        total_loss += float(loss) * batch_size

    return total_loss / total_examples

def test(loader):
    model.eval()

    correct = 0
    for data_test in loader:  # Iterate in batches over the training/test dataset.
        out = model(data_test.x_dict, data_test.edge_index_dict, data_test.batch_dict)  
        pred = out.argmax(dim=1)  # Use the class with highest probability.
        correct += int((pred == data_test.y).sum())  # Check against ground-truth labels.
    return correct / len(loader.dataset)
```

**Reply by rusty1s** [reply]
I assume that the slicing in `out[:batch_size]` and `batch.y[:batch_size]` is not really needed. Otherwise, this looks good to me.

**Reply by Zytrus** [reply]
Thank you for the feedback! I suppose I'll just have to find a more fitting model with different layers then for better results?

Also you've been a great help with my progress, thank you very much, it is very appreciated! (:

**Reply by rusty1s** [reply]
Yes, I suggest to do some design space exploration (#layers, #hidden size, LR, different global pooling operators, ..) to see if you can improve results.
